Levelled log

kjeldflarup opened the pull request:
Add UA_Log_Stdout_ErrorLevel which only prints Error level and above regardless of UA_LOGLEVEL being set lower.  

The logger can be enabled like this:
```
UA_ServerConfig *open62541_server_config = UA_Server_getConfig(open62541_server);
UA_ServerConfig_setDefault(open62541_server_config);
open62541_server_config->logger = UA_Log_Stdout_ErrorLevel_; 
```

This also allows log level being a command line parameter.
```
UA_Logger mylogger = {UA_Log_Stdout_loglevelcheck, (void*)mylevel, UA_Log_Stdout_clear};
```

The PR also makes it easier to clone UA_Log_Stdout, as some definitions is moved to `open62541/plugin/log_stdout.h`

The PR is based on 1.0.1, but should also be possible to cherry pick to master.

jpfr commented:
Good idea.
This could be generalized even further.

The logger has a void context pointer.
We could add the actual max log level to the context. That is evaluated only if a context has been set.

The existing logger can be extended that way with an alternative initialization function that takes the actual log level as an additional argument.

kjeldflarup commented:
@jpfr 
I did not change `UA_Log_Stdout_log`, because I did not want to break backwards compability, I would hate to break others peoples code.

Actually I did use context, in a very primitive way. To avoid messing with allocating variables, I typecasted it into `UA_LogLevel`.  
Now the meaning of context is defined by the `log` and clear `fields` in `UA_Logger`, thus we can continue with this typecast for now, and later change it to an allocated struct.

The idea of an initialization function is clearly better than exporting `UA_Log_Stdout_ErrorLevel`, and would support encapsulating the usage of context

kjeldflarup commented:
@jpfr
How much does size and speed matter? I assume this is embedded after all.  

I noticed that we also have `UA_LogCategory`, I would change that to a bitmask. This would allow to select categories like this  `UA_LOGCATEGORY_SERVER | UA_LOGCATEGORY_SECURITYPOLICY`

I could do the same with `UA_LogLevel`, but that does not make much sense.

Even a 16 bit void pointer for `context` could hold that
```
struct {
  level: 4;
  categorymask: 12;
};
```

kjeldflarup commented:
@jpfr I think the PR is ready for review now

jpfr commented on plugins/include/open62541/plugin/log_stdout.h:
```diff
@@ -8,9 +8,35 @@
+/* ANSI escape sequences for color output taken from here:
+ * https://stackoverflow.com/questions/3219393/stdlib-and-colored-output-in-c*/
+
+#ifdef UA_ENABLE_LOG_COLORS
```
Keep the defines in the .c file.
They are internal and not part of the API.

kjeldflarup replied:
Moved it latest commit

jpfr commented on plugins/include/open62541/plugin/log_stdout.h:
```diff
@@ -8,9 +8,35 @@
+# define ANSI_COLOR_RESET   ""
+#endif
+
+extern UA_EXPORT const char *logLevelNames[6];
```
Keep these internal as well.

kjeldflarup replied:
Moved it latest commit

jpfr commented on plugins/include/open62541/plugin/log_stdout.h:
```diff
@@ -22,6 +48,15 @@ UA_Log_Stdout_log(void *_, UA_LogLevel level, UA_LogCategory category,
+/* By default the client and server is configured with UA_Log_Stdout_log
+   These functions sets the context with a max log level */
+UA_EXPORT void
+UA_ClientConfig_setLog_Stdout_withLevel( UA_ClientConfig *config, 
```
I'd rather see a "constructor function" that returns the configured UA_Logger structure.
This can be used for the client and server config setup.

```c
UA_Logger UA_Log_Stdout_withLevel(UA_LogLevel minlevel);
```

Remove the _setLog methods.

kjeldflarup replied:
IMHO the `_setLog` is more consistent with how other parts of the client/server is configured. 

You want the users to configure it like this then?
```
conf->logger = UA_Log_Stdout_withLevel(UA_LogLevel minlevel);
```

kjeldflarup replied:
I have added the `UA_Log_Stdout_withLevel` constructor.  
But so far I have kept `UA_ClientConfig_setLog_Stdout_withLevel`, and just changed it to use `UA_Log_Stdout_withLevel`

jpfr commented on plugins/ua_log_stdout.c:
```diff
@@ -48,8 +27,14 @@ const char *logCategoryNames[7] = {"network", "channel", "session", "server",
+
+    /* Assume that context is casted to UA_LogLevel */
+    /* TODO we may later change this to a struct with bitfields to filter on category */
+    if ( context != NULL && (UA_LogLevel)context > level )
```
You might have to cast to (uintptr_t) first.
Some compilers warn for a direct cast to enums.

kjeldflarup commented:
In many plugins I see stuff like this
```UA_LOG_INFO(UA_Log_Stdout
```

Am I right that these plugins will never respect this log level?

kjeldflarup commented:
@jpfr I have updated the PR

kjeldflarup commented:
@jpfr I have added an example

jpfr commented:
I don’t like the specialization for the client and server config. The config is public API. We can have users directly set the logger and avoid too much shallow wrapper API.

Why extern the logging category names?

Please rebase on 1.1 and see that the CI is green. Can be merged when the above is done.

Good job!

kjeldflarup commented:
I have rebased and removed some unused changed.  
I have fixed the CI errors which I understood.

But the three ones left I do not understand what the problem is.

jpfr commented:
Rebased again.

jpfr commented:
Merged. Thanks for the contribution!
